"address already in use" integration test flake

```
I0409 14:17:42.800392    2593 master.go:439] Started Kubernetes API at 127.0.0.1:60623/api/v1beta1
I0409 14:17:42.800412    2593 master.go:439] Started Kubernetes API at 127.0.0.1:60623/api/v1beta2
I0409 14:17:42.800416    2593 master.go:439] Started Kubernetes API at 127.0.0.1:60623/api/v1beta3
I0409 14:17:42.800419    2593 master.go:439] Started OpenShift API at 127.0.0.1:60623/osapi/v1beta1
I0409 14:17:42.800422    2593 master.go:439] Started OAuth2 API at 127.0.0.1:60623/oauth
I0409 14:17:42.800425    2593 master.go:439] Started login server at 127.0.0.1:60623/login
I0409 14:17:42.800428    2593 master.go:439] Started OpenShift UI 127.0.0.1:60623/console/
I0409 14:17:42.800432    2593 master.go:439] Started Swagger Schema API at 127.0.0.1:60623/swaggerapi/
I0409 14:17:42.823281    2593 net.go:39] Got error &net.OpError{Op:"dial", Net:"tcp", Addr:(*net.TCPAddr)(0xc208db9440), Err:0x3d}, trying again: "127.0.0.1:60623"
I0409 14:17:42.948517    2593 master.go:496] No master policy found.  Creating bootstrap policy based on: /var/folders/zw/j6656mbn1rv4fnqqdggss8h80000gn/T/openshift-integration/policy/policy.json
I0409 14:17:43.045505    2593 factory.go:79] creating scheduler from algorithm provider 'DefaultProvider'
I0409 14:17:43.045516    2593 factory.go:129] creating scheduler with fit predicates 'map[HostName:{} PodFitsPorts:{} PodFitsResources:{} NoDiskConflict:{} MatchNodeSelector:{}]' and priority functions 'map[LeastRequestedPriority:{} ServiceSpreadingPriority:{}]
I0409 14:17:43.046122    2593 master.go:117] Started Kubernetes Scheduler
I0409 14:17:43.046134    2593 master.go:94] Started Kubernetes Replication Manager
I0409 14:17:43.046138    2593 master.go:102] Started Kubernetes Endpoint Controller
I0409 14:17:43.262383    2593 master.go:152] Started Kubernetes Minion Controller
I0409 14:17:43.262416    2593 master.go:87] Started Kubernetes Namespace Manager
I0409 14:17:43.262429    2593 start_master.go:363] Using images from "openshift/origin-<component>:"
W0409 14:17:43.262530    2593 master.go:605] Binding DNS on port 60624 instead of 53 (you may need to run as root and update your config), using 127.0.0.1:60624 which will not resolve from all locations
2015/04/09 14:17:43 skydns: ready for queries on local. for tcp://127.0.0.1:60624 [rcache 0]
2015/04/09 14:17:43 skydns: ready for queries on local. for udp://127.0.0.1:60624 [rcache 0]
I0409 14:17:43.270362    2593 net.go:39] Got error &net.OpError{Op:"dial", Net:"tcp", Addr:(*net.TCPAddr)(0xc20884b950), Err:0x3d}, trying again: "127.0.0.1:60624"
I0409 14:17:43.374807    2593 master.go:620] OpenShift DNS listening at 127.0.0.1:60624
I0409 14:17:44.317131    2593 empties.go:28] AuthenticationError: Get https://127.0.0.1:60622: dial tcp 127.0.0.1:60622: connection refused
--- FAIL: TestOAuthBasicAuthPassword (4.87s)
	oauth_basicauth_test.go:223: unexpected error: listen tcp 127.0.0.1:60622: bind: address already in use
	oauth_basicauth_test.go:271: Unexpected error: an error on the server has prevented the request from succeeding
FAIL
```

We think it has something to do with unstarted HTTP listeners racing.  Maybe.

--- a/test/integration/oauth_basicauth_test.go
+++ b/test/integration/oauth_basicauth_test.go
@@ -202,7 +202,10 @@ func TestOAuthBasicAuthPassword(t *testing.T) {
 	})
 
 	// Start remote server
-	remoteAddr := testutil.FindAvailableBindAddress()
+	remoteAddr, err := testutil.FindAvailableBindAddress(9443, 9999)
+	if err != nil {
+		t.Fatalf("Couldn't get free address for test server: %v", err)
+	}
 	remoteServer := &http.Server{
 		Addr:           remoteAddr,
 		Handler:        remoteHandler,

--- a/test/util/server.go
+++ b/test/util/server.go
@@ -1,6 +1,7 @@
 package util
 
 import (
+	"errors"
 	"fmt"
 	"net"
 	"net/url"
@@ -11,6 +12,7 @@ import (
 	kclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/golang/glog"
 
 	"github.com/openshift/origin/pkg/client"
 	newproject "github.com/openshift/origin/pkg/cmd/admin/project"
@@ -41,13 +43,26 @@ func GetBaseDir() string {
 	return cmdutil.Env("BASETMPDIR", path.Join(os.TempDir(), "openshift-"+Namespace()))
 }
 
-func FindAvailableBindAddress() string {
-	l, err := net.Listen("tcp", "127.0.0.1:0")
-	if err != nil {
-		panic(err)
+// FindAvailableBindAddress returns a bind address on 127.0.0.1 with a free port in the low-high range.
+// If lowPort is 0, an ephemeral port is allocated.
+func FindAvailableBindAddress(lowPort, highPort int) (string, error) {
+	if highPort < lowPort {
+		return "", errors.New("lowPort must be <= highPort")
 	}
-	defer l.Close()
-	return l.Addr().String()
+	for port := lowPort; port <= highPort; port++ {
+		l, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", port))
+		if err != nil {
+			if port == 0 {
+				// Only get one shot to get an ephemeral port
+				return "", err
+			}
+			continue
+		}
+		defer l.Close()
+		return l.Addr().String(), nil
+	}
+
+	return "", fmt.Errorf("Could not find available port in the range %d-%d", lowPort, highPort)
 }
 
 func setupStartOptions() (*start.MasterArgs, *start.NodeArgs, *start.ListenArg, *start.ImageFormatArgs, *start.KubeConnectionArgs) {
@@ -62,9 +77,13 @@ func setupStartOptions() (*start.MasterArgs, *start.NodeArgs, *start.ListenArg, 
 	nodeArgs.MasterCertDir = masterArgs.ConfigDir.Value()
 
 	// don't wait for nodes to come up
-	masterAddr := FindAvailableBindAddress()
-	if len(os.Getenv("OS_MASTER_ADDR")) > 0 {
-		masterAddr = os.Getenv("OS_MASTER_ADDR")
+	masterAddr := os.Getenv("OS_MASTER_ADDR")
+	if len(masterAddr) == 0 {
+		if addr, err := FindAvailableBindAddress(8443, 8999); err != nil {
+			glog.Fatalf("Couldn't find free address for master: %v", err)
+		} else {
+			masterAddr = addr
+		}
 	}
 	fmt.Printf("masterAddr: %#v\n", masterAddr)
 
@@ -72,9 +91,13 @@ func setupStartOptions() (*start.MasterArgs, *start.NodeArgs, *start.ListenArg, 
 	listenArg.ListenAddr.Set(masterAddr)
 	masterArgs.EtcdAddr.Set(GetEtcdURL())
 
-	dnsAddr := FindAvailableBindAddress()
-	if len(os.Getenv("OS_DNS_ADDR")) > 0 {
-		dnsAddr = os.Getenv("OS_DNS_ADDR")
+	dnsAddr := os.Getenv("OS_DNS_ADDR")
+	if len(dnsAddr) == 0 {
+		if addr, err := FindAvailableBindAddress(8053, 8100); err != nil {
+			glog.Fatalf("Couldn't find free address for DNS: %v", err)
+		} else {
+			dnsAddr = addr
+		}
 	}
 	fmt.Printf("dnsAddr: %#v\n", dnsAddr)
 	masterArgs.DNSBindAddr.Set(dnsAddr)

